### ISSUE TITLE
unused_declaration false positives on SceneDelegate and unrelated delegate methods.

### ISSUE THREAD

**Issue by iressler** [realm/SwiftLint#4706]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

After enabling the unused_declaration analyzer rule it reports some delegates classes or methods (e.g. `SceneDelegate`, `WKUIDelegate.webView(<args>)`) as unused, which while they aren't referenced in our code they are used by the system.

Interestingly the delegate methods only get reported as unused when they are on a subclass where the parent class does *not* implement the method. I did verify that the method is properly called on the child class, which can be seen in the sample project below.

### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint analyze --compiler-log-path <build_log_path>
Analyzing Swift files in current working directory
Collecting 'WebView.swift' (1/4)
Collecting 'AppDelegate.swift' (2/4)
Collecting 'ContentView.swift' (3/4)
Collecting 'SceneDelegate.swift' (4/4)
Analyzing 'WebView.swift' (1/4)
<path>/WebAppTest/WebAppTest/WebView.swift:26:8: error: Unused Declaration Violation: Declarations should be referenced at least once within all files linted. (unused_declaration)
Analyzing 'AppDelegate.swift' (2/4)
Analyzing 'ContentView.swift' (3/4)
Analyzing 'SceneDelegate.swift' (4/4)
<path>/WebAppTest/WebAppTest/SceneDelegate.swift:12:7: error: Unused Declaration Violation: Declarations should be referenced at least once within all files linted. (unused_declaration)
Done analyzing! Found 2 violations, 2 serious in 4 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
  0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
  Homebrew
* Paste your configuration file:

```yml
analyzer_rules:
  - unused_declaration
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  No.
* Which Xcode version are you using (check `xcodebuild -version`)?
  Xcode 14.2
  Build version 14C18
* Do you have a sample that shows the issue?
  I have attached a minimum sample project that reproduces the false positives, primarily for the WKUIDelegate method, however every project I've tried with a SceneDelegate reports it as unused.
[WebAppTest.zip](https://github.com/realm/SwiftLint/files/10458513/WebAppTest.zip)

```swift
// This triggers a violation:
class SceneDelegate: UIResponder, UIWindowSceneDelegate {
}
```
